ILC throws an IndexOutOfRangeException

### Description

I'm running into an ILC exception whenever I try to compile a project for Android when referencing the WeCantSpell.Hunspell package. I attempted to copy the source files from that repo over instead of the package reference and the same exception happened.

### Reproduction Steps

Repro:
1. Clone https://github.com/unoplatform/uno. master should be fine, but at the time of writing, f6b0f49064797ffebbde8ea78a2d913de381a051 was reproducing the problem.
2. Add a reference to WeCantSpell.Hunspell in Uno.UI.Skia.csproj [here](https://github.com/unoplatform/uno/blob/f6b0f49064797ffebbde8ea78a2d913de381a051/src/Uno.UI/Uno.UI.Skia.csproj#L35).
3. Compile the SamplesApp.Skia.netcodemobile project
```bash
dotnet publish src/SamplesApp/SamplesApp.Skia.netcoremobile/SamplesApp.Skia.netcoremobile.csproj -c Release -r android-x64 -f net10.0-android -p:UnoTargetFrameworkOverride=net10.0-android -p:ApplicationId=uno.platform.samplesapp.skia.nativeaot -p:SkiaPublishAot=true -p:SkiaSharpVersion=3.119.2-preview.1 /bl
```
4. watch it fail with the ILC exception above.

### Expected behavior

Successful build

### Actual behavior

```
EXEC : error : Index was outside the bounds of the array. [/agent/_work/1/s/src/SamplesApp/SamplesApp.Skia.netcoremobile/SamplesApp.Skia.netcoremobile.csproj::TargetFramework=net10.0-android]
  System.IndexOutOfRangeException: Index was outside the bounds of the array.
     at ILCompiler.TypePreinit.TryScanMethod(MethodIL, TypePreinit.Value[], Stack`1, Int32&, TypePreinit.Value&) + 0x6de0
     at ILCompiler.TypePreinit.TryScanMethod(MethodIL, TypePreinit.Value[], Stack`1, Int32&, TypePreinit.Value&) + 0x5801
     at ILCompiler.TypePreinit.TryScanMethod(MethodIL, TypePreinit.Value[], Stack`1, Int32&, TypePreinit.Value&) + 0x5801
     at ILCompiler.TypePreinit.TryScanMethod(MethodIL, TypePreinit.Value[], Stack`1, Int32&, TypePreinit.Value&) + 0x5801
     at ILCompiler.TypePreinit.ScanType(CompilationModuleGroup, ILProvider, TypePreinit.TypePreinitializationPolicy, ReadOnlyFieldPolicy, FlowAnnotations, MetadataType) + 0x10c
     at Internal.TypeSystem.LockFreeReaderHashtable`2.CreateValueAndEnsureValueIsInTable(TKey) + 0x14
     at ILCompiler.PreinitializationManager.IsPreinitialized(MetadataType) + 0x79
     at ILCompiler.PreinitializationManager.HasLazyStaticConstructor(TypeDesc) + 0x5d
     at ILCompiler.DependencyAnalysis.ReflectedFieldNode.GetStaticDependencies(NodeFactory) + 0x26f
     at ILCompiler.DependencyAnalysisFramework.DependencyAnalyzer`2.GetStaticDependenciesImpl(DependencyNodeCore`1) + 0x3c
     at ILCompiler.DependencyAnalysisFramework.DependencyAnalyzer`2.ProcessMarkStack() + 0xf1
     at ILCompiler.DependencyAnalysisFramework.DependencyAnalyzer`2.ComputeMarkedNodes() + 0x4d
     at ILCompiler.ILScanner.ILCompiler.IILScanner.Scan() + 0x55f
     at ILCompiler.Program.<Run>g__RunScanner|4_0(Program.<>c__DisplayClass4_0&) + 0x19c
     at ILCompiler.Program.Run() + 0x2c3d
     at ILCompiler.ILCompilerRootCommand.<>c__DisplayClass261_0.<.ctor>b__0(ParseResult) + 0x31a
/home/AzDevOps/.nuget/packages/microsoft.dotnet.ilcompiler/10.0.1/build/Microsoft.NETCore.Native.targets(330,5): error MSB3073: The command ""/home/AzDevOps/.nuget/packages/runtime.linux-x64.microsoft.dotnet.ilcompiler/10.0.1/tools/ilc" @"obj/Release/net10.0-android/android-x64/native/SamplesApp.ilc.rsp"" exited with code 1. [/agent/_work/1/s/src/SamplesApp/SamplesApp.Skia.netcoremobile/SamplesApp.Skia.netcoremobile.csproj::TargetFramework=net10.0-android]
```

This is grabbed from a CI run, see [the full run](https://dev.azure.com/uno-platform/Uno%20Platform/_build/results?buildId=193087&view=logs&jobId=531e33cc-ca50-522f-5bd0-5f3dea356f53&j=531e33cc-ca50-522f-5bd0-5f3dea356f53&t=d03a6f59-a78e-5bc7-bbf8-d8d824b969e7&s=a30de33d-9844-50aa-f54c-9171fd653fb8) for more details. If the link is dead at the time you're reading this, here's the raw log.
[log.txt](https://github.com/user-attachments/files/24980877/log.txt)

### Regression?

_No response_

### Known Workarounds

_No response_

### Configuration

This reproduces on .NET SDK 10.0.102 on x64 machines when compiling for RID=android-x64 (but might also reproduce in other setups).
This reproduces on Ubuntu images from Azure pipelines and on my Windows 11 machine (26200.7623).

### Other information

_No response_

--- a/src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/TypePreinit.cs
+++ b/src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/TypePreinit.cs
@@ -1272,7 +1272,7 @@ namespace ILCompiler
 
                         uint count = reader.ReadILUInt32();
                         int nextInstruction = reader.Offset + (int)(4 * count);
-                        if (target > count)
+                        if (target >= count)
                         {
                             reader.Seek(nextInstruction);
                         }

--- a/src/tests/nativeaot/SmokeTests/Preinitialization/Preinitialization.cs
+++ b/src/tests/nativeaot/SmokeTests/Preinitialization/Preinitialization.cs
@@ -821,6 +821,7 @@ class TestSwitch
         public static int CaseMinus1 = Switch(-1);
         public static int Case0 = Switch(0);
         public static int Case6 = Switch(6);
+        public static int Case7 = Switch(7); // Boundary: value == case count (tests fix for https://github.com/dotnet/runtime/issues/123833)
         public static int Case100 = Switch(100);
 
         private static int Switch(int x)
@@ -845,6 +846,7 @@ class TestSwitch
         Assert.AreEqual(Switcher.CaseMinus1, 100000);
         Assert.AreEqual(Switcher.Case0, 100);
         Assert.AreEqual(Switcher.Case6, 700);
+        Assert.AreEqual(Switcher.Case7, 100000);
         Assert.AreEqual(Switcher.Case100, 100000);
     }
 }

